spec/controllers/api/users_controller_spec.rb

```diff
@@ -30,4 +37,78 @@
     end
   end
 
+  describe 'upload_avatar' do
+    context 'success' do
+      it 'updates a users profile picture when uploaded' do
+        user.update avatar_kind: 'gravatar'
+        allow_any_instance_of(Paperclip::Attachment).to receive(:save).and_return(true)
+        post :upload_avatar, file: fixture_for('images', 'strongbad.png')
+        expect(response.status).to eq 200
+        expect(user.reload.avatar_kind).to eq 'uploaded'
+        expect(user.reload.uploaded_avatar).to be_present
+      end
+    end
+
+    context 'failure' do
+      it 'does not upload an invalid file' do
+        user.update avatar_kind: 'gravatar'
+        post :upload_avatar, file: fixture_for('images', 'strongbad.png', filetype: 'text/pdf')
+        expect(response.status).to_not eq 200
+        expect(user.reload.avatar_kind).to eq 'gravatar'
+        expect(user.reload.uploaded_avatar).to be_blank
+      end
+    end
+  end
+
+  describe 'change_password' do
+    context 'success' do
+      it "changes a users password" do
+        old_password = user.encrypted_password
+        post :change_password, user: { current_password: 'complex_password', password: 'new_password', password_confirmation: 'new_password'}, format: :json
+        expect(response).to be_success
+        expect(user.reload.encrypted_password).not_to eq old_password
+        json = JSON.parse(response.body)
+        user_emails = json['users'].map { |v| v['email'] }
+        expect(user_emails).to include user.email
+      end
+    end
+
+    context 'failures' do
+      it 'does not allow a change if current password does not match' do
+        old_password = user.encrypted_password
+        post :change_password, user: { current_password: 'not right', password: 'new_password', password_confirmation: 'new_password'}, format: :json
+        expect(response).to_not be_success
+        expect(user.reload.encrypted_password).to eq old_password
+      end
+
+      it 'does not allow a change if passord confirmation doesnt match' do
+        old_password = user.encrypted_password
+        post :change_password, user: { password: 'new_password', password_confirmation: 'errwhoops'}, format: :json
+        expect(response).to_not be_success
+        expect(user.reload.encrypted_password).to eq old_password
+      end
+    end
+  end
+
+  describe 'deactivate' do
```

Comment:
These API tests sure are great. It's nice that they're also a great place to confirm emails have been sent. (I realise we're not sending any in this case...)

I say that because in cucumber/rails land, we test for email sent in the cukes.. But I think it's going to be more fun for us to just test the UI with protractor, and test emails at this level.